Sparse CSR CUDA: add `torch.addmm` with all inputs sparse

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #65606
* #64391
* #61536
* #61858
* #61857
* #63948
* __->__ #63511

This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

Fixes https://github.com/pytorch/pytorch/issues/67092.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D31809838](https://our.internmc.facebook.com/intern/diff/D31809838)

IvanYashchuk committed:
Sparse CSR CUDA: add `torch.addmm` with all inputs sparse

This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -222,17 +222,22 @@ Tensor addmm_sparse_csr_dense(
+  if (result.is_sparse_csr()) {
+    zero = at::empty({mat1.size(0), mat2.size(1)}, mat2.options());
+  } else {
+    zero = at::zeros({mat1.size(0), mat2.size(1)}, mat2.options());
```
I'm trying to think of a reason as to why we're not using empty for both.

IvanYashchuk replied:
It should be zeros for both, not empty, but for sparse csr empty here is equivalent to a matrix full of zeros. `at::zeros` is not implemented for sparse csr yet.

cpuhrsch replied:
Wouldn't at::zeros for Sparse just be an entirely empty Tensor, i.e. values, crow and col indices all have 0 numel and only the size is set?

cpuhrsch commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -108,32 +112,36 @@ Tensor& addmm_out_sparse_csr_dense_cuda(
+
+  if (mat2.is_sparse_csr() && (mat1._nnz() == 0 || mat2._nnz() == 0)) {
+    if (beta.toComplexDouble() == 0.) {
+      result.values().zero_();
```
Why don't we also want to reset the column and row indices? Instead of setting the values to zero we could erase those.

IvanYashchuk replied:
I didn't think about it. The current code is simpler, erasing the sparsity pattern might not be the desired behavior. Better to do it explicitly with some other functionality.

cpuhrsch replied:
I think this entirely depends on the subsequent operation. You could make the argument that filling values with zeros is wasteful (and it indeed causes another kernel launch). Not sure what we should do here. Maybe @ngimel has some advice?

cpuhrsch commented on test/test_linalg.py:
```diff
@@ -5925,47 +5970,6 @@ def tracker(worker):
 ---(input size: {:4}, eigenpairs:{:2}, units: relative error, maxiter={:4})---
 '''.format(tol, eq_err, eq_err_general, iters1, eq_err_scipy, eq_err_general_scipy, iters2, m, k, niter))
 
-    def _test_addmm_addmv(self, f, t, m, v, *, alpha=None, beta=None, transpose_out=False, layout=torch.strided):
```
If guess if you wanted to keep this as a method of TestLinalg you could turn it into a classmethod

IvanYashchuk replied:
I want to import this function in test_sparse_csr.py

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -173,9 +187,172 @@ void addmm_out_sparse_csr(
+  } else if (mat2.is_sparse_csr() && result.is_sparse_csr()) {
+    return spgemm(mat1, mat2, beta, alpha, result);
+  } else {
+    TORCH_INTERNAL_ASSERT(false, "Received unexpected tensor layouts as input.");
```
No, it's possible to get this error only when the developer registers to call this function with the wrong dispatch key in native_functions.yaml.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch reviewed (changes requested):
Three small comments left

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented:
@IvanYashchuk - could you rebase this please?

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D31809838](https://our.internmc.facebook.com/intern/diff/D31809838)

[ghstack-poisoned]

IvanYashchuk commented:
Seems like [Windows tests](https://github.com/pytorch/pytorch/runs/3957645155) fail with a segfault... there is no error message 😕 Do not merge yet.
```
  test_addmm_all_sparse_csr_cuda_complex128 (__main__.TestSparseCSRCUDA) ... Traceback (most recent call last):
  File "run_test.py", line 1031, in <module>
    main()
  File "run_test.py", line 1009, in main
    raise RuntimeError(err_message)
RuntimeError: test_sparse_csr failed!
```

IvanYashchuk commented:
https://github.com/pytorch/pytorch/pull/59980#issuecomment-920727856
> The problem was with cuSPARSE from CUDA Toolkit on Windows segfaulting on complex128 dtype inputs for versions older than 11.2.1, it works fine with 11.2.1+. An error is raised for this case and the complex128 dtype test is skipped for older versions.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D31809838](https://our.internmc.facebook.com/intern/diff/D31809838)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D31809838](https://our.internmc.facebook.com/intern/diff/D31809838)

[ghstack-poisoned]

IvanYashchuk commented:
@cpuhrsch, I rebased and added an error for the problematic Windows CUDA older than 11.2 and complex128 input, this dtype is skipped in tests for that environment.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: add `torch.addmm` with all inputs sparse"


This PR adds `torch.addmm(c, a, b)` variant with `c, a, b` all being CSR tensors.
The underlying cuSPARSE function works only with 32-bit indices, and in
the current implementation the result tensor has 32-bit indices. Input
tensors can have both 64-bit and 32-bit indices tensors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D31809838](https://our.internmc.facebook.com/intern/diff/D31809838)

[ghstack-poisoned]
